suite/tests/client-interface/drx_scattergather-test.c

```diff
@@ -0,0 +1,772 @@
+#    define CONCAT_XMM_YMM_ZMM_U32 \
+        (XMM_REG_SIZE + YMM_REG_SIZE + ZMM_REG_SIZE) / sizeof(uint32_t)
+#    define CONCAT_XMM_YMM_U32 (XMM_REG_SIZE + YMM_REG_SIZE) / sizeof(uint32_t)
+#    define SPARSE_TEST_BUF_SIZE_U32 SPARSE_FACTOR *ZMM_REG_SIZE / sizeof(uint32_t)
+
+static bool
+test_avx512_gather(void (*test_func)(uint32_t *, uint32_t *, uint32_t *),
+                   uint32_t *ref_sparse_test_buf, uint32_t *ref_xmm_ymm_zmm,
+                   uint32_t *test_idx_vec, uint32_t *output_xmm_ymm_zmm OUT)
+{
+    memset(output_xmm_ymm_zmm, 0, CONCAT_XMM_YMM_ZMM_U32 * sizeof(uint32_t));
+    test_func(ref_sparse_test_buf, test_idx_vec, output_xmm_ymm_zmm);
+    if (memcmp(output_xmm_ymm_zmm, ref_xmm_ymm_zmm,
+               CONCAT_XMM_YMM_ZMM_U32 * sizeof(uint32_t)) != 0) {
+        print("ERROR: gather result does not match\n");
+        return false;
+    }
+    return true;
+}
+
+static bool
+test_avx2_gather(void (*test_func)(uint32_t *, uint32_t *, uint32_t *),
+                 uint32_t *ref_sparse_test_buf, uint32_t *ref_xmm_ymm,
+                 uint32_t *test_idx_vec, uint32_t *output_xmm_ymm OUT)
+{
+    memset(output_xmm_ymm, 0, CONCAT_XMM_YMM_U32 * sizeof(uint32_t));
+    test_func(ref_sparse_test_buf, test_idx_vec, output_xmm_ymm);
+    if (memcmp(output_xmm_ymm, ref_xmm_ymm, CONCAT_XMM_YMM_U32 * sizeof(uint32_t)) != 0) {
+        print("ERROR: gather result does not match\n");
+        return false;
+    }
+    return true;
+}
+
+static bool
+test_avx512_scatter(void (*test_func)(uint32_t *, uint32_t *, uint32_t *),
+                    uint32_t *ref_sparse_test_buf, uint32_t *ref_xmm_ymm_zmm,
+                    uint32_t *test_idx_vec, bool check_half, bool check_64bit_values,
+                    uint32_t *output_sparse_test_buf OUT)
+{
+    /* For scatters with maximal 8 indices, only half the sparse array is scattered. */
+    int check_size = check_half ? SPARSE_TEST_BUF_SIZE_U32 / 2 : SPARSE_TEST_BUF_SIZE_U32;
+    memset(output_sparse_test_buf, 0, SPARSE_TEST_BUF_SIZE_U32 * sizeof(uint32_t));
+    test_func(ref_xmm_ymm_zmm, test_idx_vec, output_sparse_test_buf);
+    for (int i = 0; i < check_size; i += SPARSE_FACTOR) {
+        if (check_64bit_values) {
+            if (*(uint64_t *)&output_sparse_test_buf[i] !=
+                *(uint64_t *)&ref_sparse_test_buf[i]) {
+                print("ERROR: scatter result does not match\n");
+                return false;
+            }
+        } else {
+            if (*(uint32_t *)&output_sparse_test_buf[i] !=
+                *(uint32_t *)&ref_sparse_test_buf[i]) {
+                print("ERROR: scatter result does not match\n");
+                return false;
+            }
+        }
+    }
+    return true;
+}
+
+static bool
+test_avx2_avx512_scatter_gather(void)
+{
+#    if defined(__AVX512F__) || defined(__AVX__)
+    uint32_t ref_sparse_test_buf[] = {
+        0x0, 0x1, 0xf, 0xf, 0x1, 0x2, 0xf, 0xf, 0x2, 0x3, 0xf, 0xf, 0x3, 0x4, 0xf, 0xf,
```

Comment:
Please add a comment explaining where this pattern came from and why it was chosen.

Reply:
Done.